README.md
# fast5tools 0.4
Tools for working with fast5 files (Oxford Nanopore DNA/RNA sequencing data).


Add the fast5tools dir to your PYTHONPATH.
Add fast5tools/bin to your PATH

Try:

$ git clone https://github.com/JohnUrban/fast5tools.git

$ cd fast5tools

$ FAST5TOOLS=${PWD}

$ FAST5TOOLSPATHS=${FAST5TOOLS}:${FAST5TOOLS}/bin:${FAST5TOOLS}/fast5tools

$ export PATH=${FAST5TOOLSPATHS}:$PATH

$ export PYTHONPATH=${FAST5TOOLSPATHS}:$PYTHONPATH

$ python setup.py test

This will test various fast5 files in rundata/ that span from the beginning of MAP to recently. It is actually running tests/testf5\_class.py on them. 

Scripts are available in the bin directory and may be run as command line
programs once fast5tools is installed.


All scripts can read:
- a single fast5
- many fast5s listed in a row
- a file-of-file-names (with .fofn extension) of fast5s
- a tarchive with fast5s in dirs and subdirs
- acombination of those


Fast5 information:
- fast5attributes.py

Fast5 stats:
- fast5stats.py - generate a table with any fast5 info you want. 
- fast5TableSummary.py -- in progress -- The intention was to be able to give summary stats for any table with any fast5 info in any order (so long as you can tell it the order).
- fast5standardSummary.py

Plotting from stats tables:
- fast5plot.py -- in progress. 

Plotting (and kmer counting/plotting) from fast5s directly:
- fast5ReadLengthPlots.py
- fast5_qual_histogram.py
- fast5_qual_v_pos_boxplot.py
- fast5plotEvents.py
- fast5plotRaw.py
- kmerCountsFromFastx.py


Extracting Fasta/Fastq:
- fast5Tofastx -- convert to fasta, fastq, quals, integer quals -- or just details about the seqs (similar to fast5stats)
- fast5DerivedFastxMoleculeStats.py

Extracting events, raw data, model information:
- fast5ToEvents.py 
- fast5ToModelinfo.py 
- fast5toRaw.py


De-barcoding:
- fast5_sw_alignment_viz.py
- fast5_sw_bardecoder.py


Other:
- fast5staypos.py - get position of basecall "stays" in read (e.g. as performed in analyses in biorxiv paper below)
- fast5_regex_parser.py - identify positions of a regex in reads (e.g. as performed in analyses in biorxiv paper below)





Please cite fast5tools and/or poreminion as:
-------------------------------------------
Urban, J. M., Bliss, J., Lawrence, C. E. & Gerbi, S. A. 

Sequencing ultra-long DNA molecules with the Oxford Nanopore MinION. 

bioRxiv (Cold Spring Harbor Labs Journals, 2015). doi:10.1101/019281 

http://biorxiv.org/content/early/2015/05/20/019281

This paper contains the first descriptions and uses of poreminion, which evolved into fast5tools.





changelog.txt
Version 0.4 - Bumped up on 2018-04-23
- Remaining poreminion functionalities incorporated to various fast5tool scripts in bin:
	- fast5ReadLengthPlots.py
	- fast5_qual_histogram.py
	- fast5_qual_v_pos_boxplot.py
	- fast5_regex_parser.py
	- fast5attributes.py
	- fast5staypos.py
	- kmerCountsFromFastx.py
- This also involved updating various fast5tool library scripts in fast5tools/
	- f5class.py
	- f5ops.py
	- fileListClass.py
	- helperops.py
- And adding:
	- edgeRops.py
	- plotops.py
- In addition to merging poreminion, other things were added/updated:
	- fast5toplotRaw.py can now also show segmentation of raw data (if segmented events present)
	- tarlite is now default. no tarlite is needed to turn it off. --tarlite is kept as legacy to not break pipelines.
	- added kpos and genomic kpos approaches to estimating kmer parameters from Tombo events

Version 0.3 - Bumped up on 2018-02-20 (changes commited between Oct2017-Feb20,2018)
- Updates to fast5class to support all other updates/changes and to interpret newer fast5s
- Added tar writing capabilities to scripts that output a lot of files
- Added fileListClass.py
- Added kmerParams.py for estimating kmer parameters from Tombo events
- Added fast5_tombo_extract.py for extracting Tombo events from fast5s
- Added barcodeclass.py
- Added fast5_sw_bardecoder.py
- Added fast5_sw_alignment_viz.py
- Added samGenomicWindows.py for identifying genomic sequences that underly signal (plus flanks)
- Added samclass.py
- Added samops.py
- Updated fast5plotEvents.py
- Added fast5plotRaw.py
- Added fast5toRaw.py
	

Note: More fast5tools development was relatively inactive during most of 2017 until October.

Fast5tools Version 0.0.2b - Unofficial bump up from 0.0.2 (July2016 - January 2017)
- Updates to fast5tools to support other changes/updates and/or new or problematic fast5s
- Fixes to various scripts
- Added --tarlite option for allowing tarballs full of fast5s to be operated on w/o extracting all of the files at once
- Updated tempdir naming to prevent collisions between parallel operations in same dir
- Updated fast5stats.py
- Added fast5standardSummary.py and standard table example in rundata/
- Added rundata/fastx/ and other new subdirs for various examples/tests
- Added fast5DerivedFastxMoleculeStats.py
- Added filterFast5DerivedFastx.py
- Added fxclass.py
- Added falcon-compatible fasta options (f5class, fast5tofastx, f5ops)
- Added extra rundata examples

Fast5tools Version 0.0.2 - July 6, 2016 
- Updates to fast5tools to support other changes/updates and new or problematic fast5s
- Removed tests.sh (replaced by unit tests)
- Packaged and reorganized fast5tools directory structure
- Updates to f5class to deal with problem fast5s
- Updates to fast5toEvents filenaming
- Adjusted f5class filebasenmae
- Update to fast5stats (verbose option to help debugging)
- Update fast5plot 
- Added fast5plotEvents
- Added parsedEventsClass.py (merged from squiggler, though not doing much with it at the moment)
- Added tools.py (some miscellaneous helper tools)
- Added hmm_class.py (merged from squiggler, though not doing much with it at the moment)



Fast5tools Version 0.0.1 -- Initial Commit - June 2, 2016 (Developed by JohnUrban during Jan2016-June2016)
- README
- License
- rundata tests
- testf5class.py
- test.sh
- fast5tofastx.py
- fast5toModelInfo.py
- fast5toEvents.py
- fast5stats.py
- fast5plot.py
- fast5filenames.py
- fast5TableSummary.py
- fast5SetDifference.py
- f5tableops.py
- f5class.py
- f5ops.py

### FAST5TOOLS


Fast5tools evolved from poreminion - therefore the poreminion version history is documented below.

Poreminion 0.4.4 -- July 24, 2015
Updated:
- fragsummary: 
   --> reports on molecules >50kb and >100kb now (in addition to >10kb)
   --> some print out variables needed to be fixed
- g4/regex
  --> g4/regex call on a lot of the same code. some updates to g4 broke the more general usage of regex. now fixed.
Other notes:
- since last version updates to fragstats, fragsummary, info, and main were made
- a license was added
- a citation was added to be used when citing poreminion

Poreminion 0.4.3 -- May 5, 2015
Updated:
- g4: now can count number of g-tracts inside identified g4 motif


Poreminion 0.4.2 -- April 28, 2015
Added subcommand:
- seqlen: print seqname,seqlen for chosen read types in set of fast5 files.



Poreminion 0.4.1 -- April 27, 2015
Added subcommands:
- staypos: BED file type output of where stays (move = 0) are in template and complement sequences (by mapping moves to appropriate sequence index)

Updated:
- some of the under-the-hood: e.g. a lot of functions moved to info.py from other files to help avoid import conflicts



Poreminion 0.4.0 -- April 27, 2015
Added subcommands:
- regex: look for regular expressions in minION reads
- g4: look for g4 motifs in minION reads

Updated:
- fragstats now allows collection of g4 counts in rast5 read types



Poreminion 0.3.2 -- April 23, 2015
- fragsummary
	- performs more summarizing now (more to come)
 
Poreminion 0.3.1 -- April 23, 2015
Added:
- fragsummary -- summarizes a fragstats table file -- lots of useful info. 0.3.2 will have even more info from fragstats summarized.
- fragrobust -- robust calculation -- caution: this is somewhat useless (see message that comes with it) and may be removed in next bump.
- info -- gathers a bunch of info from a given fast5 file

Updated:
- fragstats
  --> has2d and hascomp fields are now 1 and 0 instead of "True" and "False"
  --> no longer computes "robustness" as this metric is somewhat useless. For now, I kept it as its own subcommand "fragrobust". Be sure to see the message there.
  --> computes a bunch of new stuff even in non-extensive mode that do not much affect the speed (columns and metrics below):
	--> 15 = channel number molecule traversed
	--> 16 = heat sink temperature while molecule traversed
	--> 17 = num called template events (after events pruned during base-calling)

fast5tools/f5ops.py
## JOHN URBAN (2015, 2016)

import os
from random import shuffle, seed
from fast5tools.f5class import *

##
#### fast5tofastx.py,
def assert_readtype(readtype, legaloptions="tc2maM"):
    allowed = readtype[0] in legaloptions
    assert allowed
    if readtype[0] == "t" and allowed:
        readtype = "template"
    elif readtype[0] == "c" and allowed:
        readtype = "complement"
    elif readtype[0] == "2" and allowed:
        readtype = "2d"
    elif readtype[0] == "m" and allowed:
        readtype = "molecule"
    elif readtype[0] == "a" and allowed:
        readtype = "all"
    elif readtype[0] == "M" and allowed:
        readtype = "MoleQual"
    return readtype

#### fast5tofastx.py,
def details(f5, readtype):
    readstats = []
    readstats.append( f5._get_pore_info_name(readtype) )
    readstats.append( f5.get_seq_len(readtype) )
    readstats.append( f5.get_mean_qscore(readtype) )
    readstats.append( f5.get_num_events(readtype) )
    try:
        readstats.append( f5.get_num_called_events(readtype) )
    except:
        readstats.append("-")
    try:
        readstats.append( f5.get_num_skips(readtype) )
    except:
        readstats.append("-")
    try:
        readstats.append( f5.get_num_stays(readtype) )
    except:
        readstats.append("-")        
    return ("\t").join([str(e) for e in readstats])


#### fast5tofastx.py,
def get_fast5tofastx_output_fxn(outtype):
    if outtype == "fasta":
        output = fasta
    elif outtype == "fastq":
        output = fastq
    elif outtype == "qual":
        output = qual
    elif outtype == "intqual":
        output = intqual
    elif outtype == "details":
        output = details
    elif outtype == "falcon" or outtype == "oldfalcon":
        output = oldfalcon
    elif outtype == "newfalcon":
        output = newfalcon
    elif outtype == "fasta_with_abspath":
        output = fasta_with_abspath
    elif outtype == "fasta_only_abspath":
        output = fasta_only_abspath
    elif outtype == "fastq_with_abspath":
        output = fastq_with_abspath
    elif outtype == "fastq_only_abspath":
        output = fastq_only_abspath
    elif outtype == "qual_with_abspath":
        output = qual_with_abspath
    elif outtype == "qual_only_abspath":
        output = qual_only_abspath
    #
    elif outtype == "fasta_with_filename":
        output = fasta_with_filename
    elif outtype == "fasta_only_filename":
        output = fasta_only_filename
    elif outtype == "fastq_with_filename":
        output = fastq_with_filename
    elif outtype == "fastq_only_filename":
        output = fastq_only_filename
    elif outtype == "qual_with_filename":
        output = qual_with_filename
    elif outtype == "qual_only_filename":
        output = qual_only_filename
    #
    elif outtype == "fasta_readstatsname":
        output = fasta_readstatsname
    elif outtype == "fasta_readstatsname_with_abspath":
        output = fasta_readstatsname_with_abspath
    elif outtype == "fasta_readstatsname_with_filename":
        output = fasta_readstatsname_with_filename
        
    elif outtype == "fastq_readstatsname":
        output = fastq_readstatsname
    elif outtype == "fastq_readstatsname_with_abspath":
        output = fastq_readstatsname_with_abspath
    elif outtype == "fastq_readstatsname_with_filename":
        output = fastq_readstatsname_with_filename
    #
    elif outtype == "qual_readstatsname":
        output = qual_readstatsname
    elif outtype == "qual_readstatsname_with_abspath":
        output = qual_readstatsname_with_abspath
    elif outtype == "qual_readstatsname_with_filename":
        output = qual_readstatsname_with_filename
    #
    return output

def get_fast5tofastx_readtype_fxn(readtype):
    if readtype == "template":
        getread = get_template_read
    elif readtype == "complement":
        getread = get_complement_read
    elif readtype == "2d":
        getread = get_2d_read
    elif readtype == "molecule":
        getread = get_molecule_read
    elif readtype == "all":
        getread = get_all_reads
    elif readtype == "MoleQual":
        getread = get_molequal_read
    return getread

def get_fast5tofastx_fxns(outtype, readtype):
    ''' '''
    ### get outtype fxn ###
    output = get_fast5tofastx_output_fxn(outtype)
##    if args.outtype == "fasta":
##        output = fasta
##    elif args.outtype == "fastq":
##        output = fastq
##    elif args.outtype == "qual":
##        output = qual
##    elif args.outtype == "intqual":
##        output = intqual
##    elif args.outtype == "details":
##        output = details
##    elif args.outtype == "falcon" or args.outtype == "oldfalcon":
##        output = oldfalcon
##    elif args.outtype == "newfalcon":
##        output = newfalcon
##    elif args.outtype == "fasta_with_abspath":
##        output = fasta_with_abspath
##    elif args.outtype == "fasta_only_abspath":
##        output = fasta_only_abspath
##    elif args.outtype == "fastq_with_abspath":
##        output = fastq_with_abspath
##    elif args.outtype == "fastq_only_abspath":
##        output = fastq_only_abspath
##    elif args.outtype == "qual_with_abspath":
##        output = qual_with_abspath
##    elif args.outtype == "qual_only_abspath":
##        output = qual_only_abspath
##    #
##    elif args.outtype == "fasta_with_filename":
##        output = fasta_with_filename
##    elif args.outtype == "fasta_only_filename":
##        output = fasta_only_filename
##    elif args.outtype == "fastq_with_filename":
##        output = fastq_with_filename
##    elif args.outtype == "fastq_only_filename":
##        output = fastq_only_filename
##    elif args.outtype == "qual_with_filename":
##        output = qual_with_filename
##    elif args.outtype == "qual_only_filename":
##        output = qual_only_filename
##    #
##    elif args.outtype == "fasta_readstatsname":
##        output = fasta_readstatsname
##    elif args.outtype == "fasta_readstatsname_with_abspath":
##        output = fasta_readstatsname_with_abspath
##    elif args.outtype == "fasta_readstatsname_with_filename":
##        output = fasta_readstatsname_with_filename
##        
##    elif args.outtype == "fastq_readstatsname":
##        output = fastq_readstatsname
##    elif args.outtype == "fastq_readstatsname_with_abspath":
##        output = fastq_readstatsname_with_abspath
##    elif args.outtype == "fastq_readstatsname_with_filename":
##        output = fastq_readstatsname_with_filename
##        
##    elif args.outtype == "qual_readstatsname":
##        output = qual_readstatsname
##    elif args.outtype == "qual_readstatsname_with_abspath":
##        output = qual_readstatsname_with_abspath
##    elif args.outtype == "qual_readstatsname_with_filename":
##        output = qual_readstatsname_with_filename
        
    ### get readtype fxn ###
    getread = get_fast5tofastx_readtype_fxn(readtype)
##    if args.readtype == "template":
##        getread = get_template_read
##    elif args.readtype == "complement":
##        getread = get_complement_read
##    elif args.readtype == "2d":
##        getread = get_2d_read
##    elif args.readtype == "molecule":
##        getread = get_molecule_read
##    elif args.readtype == "all":
##        getread = get_all_reads
##    elif args.readtype == "MoleQual":
##        getread = get_molequal_read
    return output, getread



######################### processing functions ######
#### e.g. used in:
#### fast5tofastx.py, 
#################################################
def get_comments(request, f5, readtype, samflag=''):
    if not request:
        return False
    elif request in ('base_info', 'pore_info', 'read_stats', 'event_stats', 'read_event_stats', 'abspath', 'filename'):
        if request == 'base_info':
            return samflag + f5.get_base_info_name()
        elif request == 'pore_info':
            return samflag + f5.get_pore_info_name(readtype)
        elif request == 'read_stats':
            return samflag + f5.get_read_stats_name(readtype)
        elif request == 'event_stats':
            return samflag + f5.get_event_stats_name(readtype)
        elif request == 'read_event_stats':
            return samflag + f5.get_read_and_event_stats_name(readtype)
        elif request == 'abspath':
            return samflag + f5.abspath
        elif request == 'filename':
            return samflag + f5.filebasename
    else:
        return samflag + request
    
def get_single_read(f5, readtype, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    ''' f5 is Fast5 object.
        readtype in template, complement, 2d, molecule, all.
        minlen/maxlen - integers.
        minq/maxq - floats
        output is a function'''
    if f5.has_read(readtype):
        if f5.get_seq_len(readtype) >= minlen and f5.get_seq_len(readtype) <= maxlen:
            if f5.get_mean_qscore(readtype) >= minq and f5.get_mean_qscore(readtype) <= maxq:
                if kwargs['comments']:
                    kwargs['comments'] = get_comments(kwargs['comments'], f5, readtype, kwargs['samflag'])
                return output(f5, readtype, *args, **kwargs)

def get_template_read(f5, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    return get_single_read(f5, "template", minlen, maxlen, minq, maxq, output, *args, **kwargs)

def get_complement_read(f5, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    return get_single_read(f5, "complement", minlen, maxlen, minq, maxq, output, *args, **kwargs)

def get_2d_read(f5, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    return get_single_read(f5, "2d", minlen, maxlen, minq, maxq, output, *args, **kwargs)

def get_molecule_read(f5, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    return get_single_read(f5, f5.use_molecule(), minlen, maxlen, minq, maxq, output, *args, **kwargs)

def get_molequal_read(f5, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    return get_single_read(f5, f5.use_molequal(), minlen, maxlen, minq, maxq, output, *args, **kwargs)


def get_all_reads(f5, minlen, maxlen, minq, maxq, output, *args, **kwargs):
    allreads = ""
    for readtype in ["template", "complement", "2d"]:
        try:
            allreads += get_single_read(f5, readtype, minlen, maxlen, minq, maxq, output, *args, **kwargs) + '\n'
        except:
            pass
    return allreads.rstrip()



######################### output functions ######
#### e.g. used in get_single_read()
#################################################


def fastq(f5, readtype, *args, **kwargs):
    return f5.get_fastq(readtype, comments=kwargs['comments'])

def fastq_with_abspath(f5, readtype, *args, **kwargs):
    return f5.get_fastq_with_abspath(readtype, comments=kwargs['comments'])

def fastq_only_abspath(f5, readtype, *args, **kwargs):
    return f5.get_fastq_only_abspath(readtype, comments=kwargs['comments'])

def fastq_with_filename(f5, readtype, *args, **kwargs):
    return f5.get_fastq_with_filename(readtype, comments=kwargs['comments'])

def fastq_only_filename(f5, readtype, *args, **kwargs):
    return f5.get_fastq_only_filename(readtype, comments=kwargs['comments'])

def fastq_readstatsname(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name(readtype)
    return f5.get_fastq(readtype, name = name, comments=kwargs['comments'])

def fastq_readstatsname_with_filename(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name_with_filebasename(readtype)
    return f5.get_fastq(readtype, name = name, comments=kwargs['comments'])

def fastq_readstatsname_with_abspath(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name_with_abspath(readtype)
    return f5.get_fastq(readtype, name = name, comments=kwargs['comments'])




def fasta(f5, readtype, *args, **kwargs):
    return f5.get_fasta(readtype, comments=kwargs['comments'])

def fasta_with_abspath(f5, readtype, *args, **kwargs):
    return f5.get_fasta_with_abspath(readtype, comments=kwargs['comments'])

def fasta_only_abspath(f5, readtype, *args, **kwargs):
    return f5.get_fasta_only_abspath(readtype, comments=kwargs['comments'])

def fasta_with_filename(f5, readtype, *args, **kwargs):
    return f5.get_fasta_with_filename(readtype, comments=kwargs['comments'])

def fasta_only_filename(f5, readtype, *args, **kwargs):
    return f5.get_fasta_only_filename(readtype, comments=kwargs['comments'])

def fasta_readstatsname(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name(readtype)
    return f5.get_fasta(readtype, name = name, comments=kwargs['comments'])

def fasta_readstatsname_with_filename(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name_with_filebasename(readtype)
    return f5.get_fasta(readtype, name = name, comments=kwargs['comments'])

def fasta_readstatsname_with_abspath(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name_with_abspath(readtype)
    return f5.get_fasta(readtype, name = name, comments=kwargs['comments'])



def qual(f5, readtype, *args, **kwargs):
    return f5.get_quals(readtype, comments=kwargs['comments'])

def qual_with_abspath(f5, readtype, *args, **kwargs):
    return f5.get_quals_with_abspath(readtype, comments=kwargs['comments'])

def qual_only_abspath(f5, readtype, *args, **kwargs):
    return f5.get_quals_only_abspath(readtype, comments=kwargs['comments'])

def qual_with_filename(f5, readtype, *args, **kwargs):
    return f5.get_quals_with_filename(readtype, comments=kwargs['comments'])

def qual_only_filename(f5, readtype, *args, **kwargs):
    return f5.get_quals_only_filename(readtype, comments=kwargs['comments'])


def qual_readstatsname(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name(readtype)
    return f5.get_quals(readtype, name = name, comments=kwargs['comments'])

def qual_readstatsname_with_filename(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name_with_filebasename(readtype)
    return f5.get_quals(readtype, name = name, comments=kwargs['comments'])

def qual_readstatsname_with_abspath(f5, readtype, *args, **kwargs):
    name = f5.get_read_stats_name_with_abspath(readtype)
    return f5.get_quals(readtype, name = name, comments=kwargs['comments'])


def intqual(f5, readtype, *args, **kwargs):
    return f5.get_quals_as_int(readtype)

def oldfalcon(f5, readtype, *args, **kwargs):
    return f5.get_falcon_fasta(readtype, zmw_num=kwargs['falcon_i'], style="old")

def newfalcon(f5, readtype, *args, **kwargs):
    return f5.get_falcon_fasta(readtype, zmw_num=kwargs['falcon_i'], style="new")

#######
f5fxn = {}
f5fxn[1] = lambda f5: f5.get_base_info_name()
f5fxn[2] = lambda f5: f5.get_seq_len(f5.use_molecule()) if f5.has_reads() else "-"
f5fxn[3] = lambda f5: f5.get_mean_qscore(f5.use_molecule()) if f5.has_reads() else "-"
f5fxn[4] = lambda f5: 1 if f5.has_read("complement") else 0
f5fxn[5] = lambda f5: 1 if f5.has_read("2d") else 0
f5fxn[6] = lambda f5: f5.get_seq_len("2d") if f5.has_read("2d") else "-"
f5fxn[7] = lambda f5: f5.get_seq_len("template") if f5.has_read("template") else "-"
f5fxn[8] = lambda f5: f5.get_seq_len("complement") if f5.has_read("complement") else "-"
f5fxn[9] = lambda f5: f5.get_mean_qscore("2d") if f5.has_read("2d") else "-"
f5fxn[10] = lambda f5: f5.get_mean_qscore("template") if f5.has_read("template") else "-"
f5fxn[11] = lambda f5: f5.get_mean_qscore("complement") if f5.has_read("complement") else "-"
f5fxn[12] = lambda f5: f5.get_num_events("input")
f5fxn[13] = lambda f5: f5.get_num_events("template") if f5.has_read("template") else "-"
f5fxn[14] = lambda f5: f5.get_num_events("complement") if f5.has_read("complement") else "-"
f5fxn[15] = lambda f5: f5.get_num_called_events("template") if f5.has_read("template") else "-"
f5fxn[16] = lambda f5: f5.get_num_called_events("complement") if f5.has_read("complement") else "-"
f5fxn[17] = lambda f5: f5.get_num_skips("template") if f5.has_read("template") else "-"
f5fxn[18] = lambda f5: f5.get_num_skips("complement") if f5.has_read("complement") else "-"
f5fxn[19] = lambda f5: f5.filename
f5fxn[20] = lambda f5: f5.abspath
f5fxn[21] = lambda f5: f5.get_log()
f5fxn[22] = lambda f5: f5.get_log_string()
##f5fxn[19] = lambda f5
##f5fxn[20] = lambda f5
##f5fxn[21] = lambda f5
##f5fxn[22] = lambda f5
##f5fxn[23] = lambda f5
##f5fxn[24] = lambda f5



## F5 FILTERING/SELECTION OPERATIONS
def define_read_type(f5, readtype):
        if readtype in ('molecule'):
            return f5.use_molecule()
        else:
            return readtype

def meets_all_criteria(f5, readtype, minlen, maxlen, minq, maxq):
    if f5.is_not_corrupt() and f5.is_nonempty():
        readtype = define_read_type(f5, readtype)
        if f5.has_read(readtype):
            if f5.get_seq_len(readtype) >= minlen and f5.get_seq_len(readtype) <= maxlen:
                if f5.get_mean_qscore(readtype) >= minq and f5.get_mean_qscore(readtype) <= maxq:
                    return True
    else:
        return False


## Fast5List OPs
## The function below allows one to sample from the Fast5List
## I just tucked this inside Fast5List class though, so it may not be needed.
## Also I did away with pre-shuffling and selection as that can create unintended weighting schemes when sampling
## get_fast5_list(args.nfiles, args.fast5, args.random, args.randomseed, args.notarlite, filemode='r', sort=True)
def get_fast5_list(nfiles, initial_list, random=False, randomseed=False, notarlite=False, filemode='r', sort=True):
    # Expand targets to get initial list
    f5list = Fast5List(initial_list, keep_tar_footprint_small=(not notarlite), filemode=filemode)
    f5list.down_sample_iter_files(n=nfiles, random=random, randomseed=randomseed, sort=True)

    ## Take from expanded list: Shuffling and Downsampling actually happens here on indiv fast5s
    #return f5list.get_sample(n=nfiles, random=random, sort=sort)
    return f5list






